File: backend/web/routers/performance.py
"""Performance API router."""

from __future__ import annotations

from fastapi import APIRouter, Request
from fastapi.responses import JSONResponse

from opencmo import storage
from opencmo.tools.performance_tracker import (
    add_manual_tracking,
    collect_approval_metrics,
    delete_manual_tracking,
    get_project_performance,
    list_manual_tracking,
)

router = APIRouter(prefix="/api/v1")


@router.get("/projects/{project_id}/performance")
async def api_v1_performance(project_id: int):
    """Get all published content performance for a project."""
    project = await storage.get_project(project_id)
    if not project:
        return JSONResponse({"error": "Not found"}, status_code=404)

    approvals = await get_project_performance(project_id)
    manual = await list_manual_tracking(project_id)

    # Compute summary stats
    total_likes = 0
    total_comments = 0
    total_retweets = 0
    for item in approvals:
        m = item.get("post_metrics") or {}
        total_likes += m.get("score", 0) + m.get("like_count", 0)
        total_comments += m.get("num_comments", 0) + m.get("reply_count", 0)
        total_retweets += m.get("retweet_count", 0)

    return JSONResponse({
        "summary": {
            "total_published": len(approvals),
            "total_likes": total_likes,
            "total_comments": total_comments,
            "total_retweets": total_retweets,
        },
        "approvals": approvals,
        "manual": manual,
    })


@router.post("/approvals/{approval_id}/refresh-metrics")
async def api_v1_refresh_metrics(approval_id: int):
    """Manually trigger metrics refresh for a published approval."""
    metrics = await collect_approval_metrics(approval_id)
    if metrics is None:
        return JSONResponse({"error": "Could not fetch metrics"}, status_code=400)
    return JSONResponse({"ok": True, "metrics": metrics})


@router.post("/projects/{project_id}/manual-tracking")
async def api_v1_add_manual_tracking(project_id: int, request: Request):
    """Add a manually tracked external URL."""
    project = await storage.get_project(project_id)
    if not project:
        return JSONResponse({"error": "Project not found"}, status_code=404)
    body = await request.json()
    url = str(body.get("url", "")).strip()
    if not url:
        return JSONResponse({"error": "url is required"}, status_code=400)
    tracking_id = await add_manual_tracking(
        project_id,
        platform=str(body.get("platform", "other")).strip(),
        url=url,
        title=str(body.get("title", "")).strip(),
        notes=str(body.get("notes", "")).strip(),
    )
    return JSONResponse({"ok": True, "id": tracking_id}, status_code=201)


@router.delete("/manual-tracking/{tracking_id}")
async def api_v1_delete_manual_tracking(tracking_id: int):
    """Delete a manually tracked item."""
    ok = await delete_manual_tracking(tracking_id)
    if not ok:
        return JSONResponse({"error": "Not found"}, status_code=404)
    return JSONResponse({"ok": True})

File: frontend/src/api/performance.ts
import { apiJson } from "./client";

export interface PerformanceItem {
  id: number;
  channel: string;
  approval_type: string;
  title: string;
  content_preview: string;
  publish_result: Record<string, unknown> | null;
  pre_metrics: Record<string, number>;
  post_metrics: Record<string, number>;
  created_at: string;
  decided_at: string;
}

export interface ManualTrackingItem {
  id: number;
  platform: string;
  url: string;
  title: string;
  notes: string;
  metrics: Record<string, number>;
  created_at: string;
}

export interface PerformanceSummary {
  total_published: number;
  total_likes: number;
  total_comments: number;
  total_retweets: number;
}

export interface PerformanceData {
  summary: PerformanceSummary;
  approvals: PerformanceItem[];
  manual: ManualTrackingItem[];
}

export function getPerformance(projectId: number): Promise<PerformanceData> {
  return apiJson<PerformanceData>(`/projects/${projectId}/performance`);
}

export function refreshMetrics(approvalId: number): Promise<{ ok: boolean; metrics?: Record<string, number> }> {
  return apiJson(`/approvals/${approvalId}/refresh-metrics`, { method: "POST" });
}

export function addManualTracking(
  projectId: number,
  data: { platform: string; url: string; title?: string; notes?: string },
): Promise<{ ok: boolean; id: number }> {
  return apiJson(`/projects/${projectId}/manual-tracking`, {
    method: "POST",
    body: JSON.stringify(data),
  });
}

export function deleteManualTracking(trackingId: number): Promise<{ ok: boolean }> {
  return apiJson(`/manual-tracking/${trackingId}`, { method: "DELETE" });
}
